ogr/ogrsf_frmts/miramon/mm_gdal_functions.c

```diff
@@ -0,0 +1,2801 @@
+            {
+                for (offset_fals = (offset_primera_fitxa - 1);
+                     !((offset_fals - 1) % 32); offset_fals--)
+                    ;
+            }
+        }
+    }
+    else
+        offset_reintent = ftell_function(pf);
+
+    if (1 != fread_function(&ushort, 2, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->reserved_1), 2, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->transaction_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->encryption_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->dbf_on_a_LAN), 12, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+    {
+        memcpy(&pMMBDXP->nRecords, &nRecords, 4);
+        memcpy(((char *)&pMMBDXP->nRecords) + 4, &pMMBDXP->dbf_on_a_LAN, 4);
+    }
+    else
+        pMMBDXP->nRecords = nRecords;
+
+    if (1 != fread_function(&(pMMBDXP->MDX_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->CharSet), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->reserved_2), 2, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    // Checking for a cpg file
+    if (pMMBDXP->CharSet == 0)
+    {
+        FILE_TYPE *f_cpg;
+
+        strcpy(cpg_file, pMMBDXP->szFileName);
+        strcpy(cpg_file, reset_extension(cpg_file, "cpg"));
+        f_cpg = fopen_function(cpg_file, "r");
+        if (f_cpg)
+        {
+            char *p;
+            size_t read_bytes;
+            fseek_function(f_cpg, 0L, SEEK_SET);
+            if (11 > (read_bytes = fread_function(local_message, 1, 10, f_cpg)))
+            {
+                local_message[read_bytes] = '\0';
+                p = strstr(local_message, "UTF-8");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_UTF8_DBF;
+                p = strstr(local_message, "UTF8");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_UTF8_DBF;
+                p = strstr(local_message, "ISO-8859-1");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_ANSI_DBASE;
+            }
+            fclose_function(f_cpg);
+        }
+    }
+    if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+    {
+        memcpy(&pMMBDXP->FirstRecordOffset, &offset_primera_fitxa, 2);
+        memcpy(((char *)&pMMBDXP->FirstRecordOffset) + 2, &pMMBDXP->reserved_2,
+               2);
+
+        if (n_queixes_estructura_incorrecta > 0)
+            offset_fals = pMMBDXP->FirstRecordOffset;
+
+        memcpy(&pMMBDXP->BytesPerRecord, &ushort, 2);
+        memcpy(((char *)&pMMBDXP->BytesPerRecord) + 2, &pMMBDXP->reserved_1, 2);
+    }
+    else
+    {
+        pMMBDXP->FirstRecordOffset = offset_primera_fitxa;
+        pMMBDXP->BytesPerRecord = ushort;
+    }
+
+    /* ====== Record structure ========================= */
+
+    if (n_queixes_estructura_incorrecta > 0)
+        pMMBDXP->nFields = (MM_EXT_DBF_N_FIELDS)(((offset_fals - 1) - 32) / 32);
```

Comment:
I suspect things might go wrong if offset_fals is 0. The resulting nFields would be likely close to UINT32_MAX / 32.

Reply:
It's a solution for bad formed DBF that perhaps in the past we adopted. To be sure it's not a mess I protect it:

if (some_problems_when_reading > 0)
    {
        if((offset_fals - 1) - 32 < 0)
            pMMBDXP->nFields = 0;
        else
            pMMBDXP->nFields = (MM_EXT_DBF_N_FIELDS)(((offset_fals - 1) - 32) / 32);
    }